File: datastructures/arrays/java/algorithms/DuplicateZeros.java
public class DuplicateZeros {

    public static void duplicateZeros(int[] arr) {
        int numberOfZeros = findNumberOfZeros(arr);
        int n = arr.length;

        for (int i = n - 1; i >= 0; i--) {
            insert(arr, i, numberOfZeros);

            if (arr[i] == 0) {
                numberOfZeros -= 1;

                insert(arr, i, numberOfZeros);
            }
        }
    }

    private static int findNumberOfZeros(int[] arr) {
        int numberOfZeros = 0;

        for (int i = 0; i < arr.length; i++) {
            if (arr[i] == 0) {
                numberOfZeros += 1;
            }
        }

        return numberOfZeros;
    }

    private static void insert(int[] arr, int i, int numberOfZeros) {
        if (i + numberOfZeros < arr.length) {
            arr[i + numberOfZeros] = arr[i];
        }
    }

    public static void main(String[] args) {
        int[] arr = { 1, 0, 2, 3, 0, 4, 5, 0 };
        duplicateZeros(arr);

        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
    }

}

File: datastructures/arrays/python/algorithms/duplicate_zeros.py
def duplicateZeros(arr):
    length = len(arr)
    zeros = countZeros(arr)

    print(f"zeros = {zeros}")

    for index in reversed(range(len(arr))):
        insert(arr, zeros, index)

        if arr[index] == 0:
            zeros -= 1
            insert(arr, zeros, index)


def countZeros(arr):
    zeros = 0
    for index in range(len(arr)):
        if arr[index] == 0:
            zeros += 1
    return zeros


def insert(arr, number_of_zeros, index):
    if (index + number_of_zeros < len(arr)):
        arr[index + number_of_zeros] = arr[index]


def main(arr):
    duplicateZeros(arr)
    for index in range(len(arr)):
        print(arr[index])


if __name__ == "__main__":
    arr = [1, 0, 2, 3, 0, 4, 5, 6]
    main(arr)
